### PR TITLE
Modifying cartesian product to allow for >2D input arrays

### PR DESCRIPTION
**Pull request by ckrapu**
Currently, using the implementation of MarginalKron for structured Gaussian processes is limited by an inability to initialize the GP for input arrays for the Kronecker product which have more than 1 dimension. For example, creating a Kronecker GP with a 2D spatial covariance kernel added to a 1D temporal covariance kernel is not possible while adding a 1D spatial to a 1D temporal kernel is possible. This issue limits the utility of the Kronecker GP from being used in arguably its most suitable use case (independent space/time covariance kernels). To recreate this issue, see [this gist](https://gist.github.com/ckrapu/1d7e2e76ce3f228b19029e4fecc9acbf).

The main change to the function `cartesian` is to first create a cartesian product of integer vectors, and then use these to index into the input arrays instead.

### CODE REVIEW

**Comment by ckrapu** [comment]
I didn't run the tests locally here - I'll edit and fix as appropriate so that it passes.

**Comment by bwengals** [comment]
I checked out the branch and tried the gist, it works great.

> arguably its most suitable use case  

Definitely agree.  

Maybe a test of the new functionality with a 2D X input could be added, or existing test modified?

**Comment by ckrapu** [comment]
Thanks for taking a look at this! I just added a new test which covers the usage of a 2D input as well as a 1D length-1 input.

File: pymc3/tests/test_math.py [hunk 1]
```diff
@@ -74,6 +74,23 @@ def test_cartesian():
...
+        [
+            [1, 2, 5, 0],
+            [1, 2, 6, 0],
+            [3, 4, 5, 0],
+            [3, 4, 6, 0],
+        ]
+    )
+    auto_cart = cartesian(a, b, c)
+    np.testing.assert_array_almost_equal(manual_cartesian, auto_cart)
+
```
Comment (bwengals) [inline pymc3/tests/test_math.py:92]: I would think that `assert_array_equal` is sufficient here (no almost).  I noticed the other `test_cartesian` does this so they should both be fixed

**Review by bwengals — CHANGES_REQUESTED** [review]

File: pymc3/tests/test_math.py (hunk 1 above)
Comment (ckrapu) [reply pymc3/tests/test_math.py:92]: Good point - those assertions have been changed.

**Review by bwengals — APPROVED** [review]

**Review by bwengals — APPROVED** [review]

**Review by michaelosthege — APPROVED** [review]
LGTM

From your example gist my understanding is that the function you changed plays a role for the `kron_Xs` in the marginal likelihood?

**Review by michaelosthege — CHANGES_REQUESTED** [review]
Ah sorry one more thing: Please add a line to the release notes.

**Review by michaelosthege — APPROVED** [review]

**Comment by ckrapu** [comment]
Thanks for adding the release note!

> LGTM
> 
> From your example gist my understanding is that the function you changed plays a role for the `kron_Xs` in the marginal likelihood?

Yes, it turns out that most of the functionality in the marginal likelihood bit was working fine - the only change was in how the input coordinates are  preprocessed via `pm.math.cartesian` from a short array of per-axis grid coordinates to a tall array of points.

**Pull request merged**